### ISSUE TITLE
[API Proposal]: Generic Text API's

### ISSUE THREAD

**Issue by Entomy** [dotnet/runtime#57980]
### Background and motivation

Every time I suggest anything it's immediately rejected, even though those ideas sometimes later come to fruition. So let me get a few points out first:

I'll design the entire API.

I'll do all the documentation.

I'll do all the testing.

I'll do all the additions.

I'll make any changes necessary after discussion in the review process.

I'll do the whole job. I just need to know I'm not wasting my time up front, since history says it'll just be rejected.

I'm not asking you to design this. I've always been comfortable doing the work and maintenance. What I want is for once to open up an actual discussion and begin some much needed work. I've been developing and maintaining a large and growing set of libraries that need this since late 2018. Additionally maintaining this feature would actually make my overall load lighter, and is in my best interest.

We've both converged on the same idea, as shown with work undergone with [Generic Math](https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/) for numerics and my own [Traits](https://github.com/Entomy/LibLangly/tree/master/Philosoft.Traits) and [Concepts](https://github.com/Entomy/LibLangly/tree/master/Philosoft.Concepts) for data structures and streams. It's a phenomenal tool for maintainability, orthogonality, and feature robustness, as I've been able to cut my own codebase down nearly a third, despite providing the same features.

But there's a major pain point I've had: Text types.

In some situations, I have over 200 overloads of what is conceptually the same operation. Given a sophisticated generics system this would be resolvable down to 1-4 implementations. I specifically chose .NET back in 2017 for text processing instead of Ada because, despite my very broad experience with it , because of the immense mess of textual types Ada had.

`Character`, `Wide_Character`, `Wide_Wide_Character`, `String`, `Wide_String`, `Wide_Wide_String`, `Bounded_String`, `Bounded_Wide_String`, `Bounded_Wide_Wide_String`, `Superbounded_String`, `Superbounded_Wide_String`, `Superbounded_Wide_Wide_String`, `Unbounded_String`, `Unbounded_Wide_String`, `Unbounded_Wide_Wide_String`... it's a mess... and that's _only_ what's in the standard library. Despite how sophisticated Ada's generic system was; having in the mid 80's what .NET is only recently getting for numerics, it never solved the text issue. And, recently, they even suggested adding _another_ text type that I managed to talk the designers out of.

The .NET languages are heading this way if this isn't resolved. `Char`, `Rune`, `String`, `Char[]`, `ArraySegment<Char>`, `Memory<Char>`, `ReadOnlyMemory<Char>`, `Span<Char>`, `ReadOnlySpan<Char>`, `IEnumerable<Char>`... And then you've got proposals for new textual types with [`Utf8String`](https://github.com/dotnet/corefxlab/issues/2350)... If that goes through the maintainance workload will double.

I refused to use Ada for text processing despite it's _excellent_ capabilities for embedded systems and other high-reliability systems, because it was careless with its design. .NET is headed this way. But I also have enough experience with it to recognize this and address it. At the time I started this on .NET, the situation wasn't anywhere near as convoluted. There was a nice, simple, clear, picture of what should be done. It's not that way anymore.

Add in the need for third-party libraries for things like working with grapheme clusters and the situation grows increasingly more complex. And you _have_ to rely on these third-party libraries for anything that isn't terribly broken.

What I call psuedotraits and what you call Generic <X> is clearly the new design pattern for addressing these kinds of situations, and I beleive can be applied in this instance.

If you agree, I'll go through and put together an API proposal. If not, I think I need to seriously look at getting my libraries off .NET. This isn't my job, and the insane amount of maintainance burden imposed on me just isn't worth it.

### API Proposal

Will provide if team doesn't immediately reject.

### API Usage

Will provide if team doesn't immediately reject.

### Risks

Will provide if team doesn't immediately reject.

**Comment by stephentoub** [comment]
@Entomy, thanks for your interest in participating. It's not clear what you're proposing, though.  No one here is capable of judging whether something will ultimately be accepted or declined before knowing what that thing actually is, and your monologue doesn't actually say.  I also read your missive as suggesting lack of good faith on the part of others involved here in the past; I don't believe that to have been the case, but regardless, we expect everyone here to participate according to the [code of conduct](https://github.com/dotnet/runtime/blob/main/CODE-OF-CONDUCT.md).  This is a collaborative space; it's rare for something to be proposed and merged exactly as it was proposed, with lots of folks sharing opinions and building on others' ideas and weighing options and jointly finding something that solves an agreed upon set of problems.  We also have a well-defined [process](https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md) for adding new APIs to the platform, with the API review team ultimately responsible for deciding what is in and the shape of what's in, and with the repo's owners ultimately responsible for deciding what's merged.  If you're uncomfortable with that and want to own a solution from ideation all the way through to finished bits, I can certainly understand that, and it sounds like you already have a good avenue for achieving that with the libraries you own and ship.  If instead you're interested in a well-intentioned collaboration, we're here and are interested in discussing your ideas.

**Comment by Entomy** [comment]
> I'll make any changes necessary after discussion in the review process.

I said the opposite of what you're accusing me of.

But at least I now have an answer to why things go so very differently with this team than elsewhere.

For over a decade things were quite good. Information comes out about a condition I have, information at least half the team had provably come into contact with. Then all of a sudden things change. While my interactions with the greater .NET community are largely the same, the interactions with this team, and two others, rapidly became different. Took me a while to understand how, but it's become clear. Take a condition well known to have trouble navigating complex social situations and inject business politics. Of course things are going to go awry for me. Since this didn't happen until _after_ it's been known what I have, well... the picture is pretty clear to me. I see @terrajobst in the likes for the above post; you know exactly what I have 😉

CoC me if you want. I have no interest in working with people who intentionally exploit disabilities.

**Comment by stephentoub** [comment]
> I said the opposite of what you're accusing me of.

I didn't accuse you of anything, and I don't appreciate your insinuations. You wrote line after line about how you wanted this to be a solo effort ("I'll do the whole job", "I'm not asking you to design this"), where you would be responsible for this whole thing end to end,  and where "this" wasn't even defined because you refused to share details until after the unshared proposal was agreed to be accepted ("If you agree"). I don't know what kind of response you expected, but I outlined our processes and our interest in hearing your ideas if you were interested in sharing them.  I'm sorry you're not.  Best wishes.

**Comment by terrajobst** [comment]
@Entomy 

> I have no interest in working with people who intentionally exploit disabilities.

I believe this is a welcoming place for developers of all creeds. But it's a two-way street. Almost every thread you're opening, you seem eager to weave in accusations without being specific; this makes it impossible to resolve any of the issues or concerns you seem to have.

> I see @terrajobst in the likes for the above post; you know exactly what I have 😉

Yes, we have talked in the past. But then you blocked me on Twitter before explicitly naming me in a thread and making factually incorrect statements about the nature of our interactions. Also, you referred to me and another coworker with the "f"-word. In the past, I've tried to go the extra mile and make things right for you, but it seems to me as if you seek an adversarial relationship, as evidenced by the wording and tone of virtually all the threads you're participating in.

It seems neither of us has a desire to work with the other side at this point, so I suggest we simply part ways. However, if you keep opening issues like the one here, we'll eventually be forced to block your account as your tone and vague accusations aren't just distractions anymore, they are bordering on harassment and thus violate the [code of conduct](https://dotnetfoundation.org/about/code-of-conduct).
